capstone/src/main/java/com/google/sps/servlets/GroupMemberServlet.java

```diff
@@ -26,24 +33,63 @@
   @Override
   // Adds a new member to a group 
   public void doPost(String userId, HttpServletRequest request, HttpServletResponse response) throws IOException {
-    
+
+    String email = request.getParameter("email");
     Long groupId = Long.parseLong(request.getParameter("groupId"));
-    
+
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity memberEntity = getMemberEntity(email, response, datastore);
     Entity group = this.getGroupFromId(response, groupId, datastore);
-    if (group != null && !userId.equals("")) {
+    if (group == null) return;
+
+    String memResponse = doPost_helper(response, datastore, groupId, memberEntity, group);
+  }
+
+  // If no account for email, return 'This email doesn't have an account'
+  // If user already in group, return 'User already in group'
+  // If user not it group, add user and return 'User added to group'
+  private String doPost_helper(HttpServletResponse response, DatastoreService datastore, Long groupId, Entity memberEntity, Entity group) {
+    String memResponse = "";
+    if (memberEntity == null) {
+      memResponse += "This email doesn't have an account.";
+    } else {
       ArrayList<String> members = 
-        (ArrayList<String>) group.getProperty("members");
+        (ArrayList<String>) group.getProperty("memberIds");
       if (members == null) {
         members = new ArrayList<>();
       }
 
-      this.addMember(members, userId);
-      group.setProperty("members", members);
-      datastore.put(group);
-    } else {
-      return;
+      String memberId = (String) memberEntity.getProperty("userId");
+      if (members.contains(memberId)) {
+        memResponse += "User is already in group.";
+      } else {
+        this.addMember(members, memberId);
+        group.setProperty("memberIds", members);
+        datastore.put(group);
+        memResponse += "User added to group.";
+        addGroupToUser(response, datastore, groupId, memberEntity);
+      }
     }
+    return memResponse;
+  }
+
+  private void addGroupToUser(HttpServletResponse response, DatastoreService datastore, Long groupId, Entity memberEntity) {
```

Comment:
Can you restructure this to return true if a groupId is successfully added and false if one is already found?

then you can just do

```java
if (maybeAddGroupToUser(...)) {
  memResponse += "User added to group."
} else {
  memResponse += "User is already in group."
}
```

in doPost_helper and I won't lose sleep over whether addGroupToUser is ever going to accidentally be used to add a second instance of the groupId to the the user 😅